### PR TITLE
Use specialized PostgreSQL types for some more columns

### LINKED ISSUE #70: Use SQLAlchemy machinery for field that are JSON encoded
SQLAlchemy has a machinery that can automatically manage fields that has to be JSON encoded on the database:

http://docs.sqlalchemy.org/en/rel_0_7/core/types.html#marshal-json-strings

We may be interested in using it.

### PR DESCRIPTION
**Pull request by lerks**
Fixes #70 and progresses #87.

Postgres has both the JSON and the JSONB types. The first one just adds validation but stores the textual representation. The second one converts to an optimized format for storage (which causes duplicate keys and the order of keys to get lost). I think we want JSONB.

---
This change is [Reviewable](https://reviewable.io/reviews/cms-dev/cms/741)

### CODE REVIEW

**Comment by lerks** [comment]
I rebased this on the current master and added a few fixes. The reason it still fails on Travis is that it introduces the use of the JSONB data type in PostgreSQL, which was added in version 9.4. Ubuntu 16.04 (the version we target) has version 9.5 and therefore should work fine. Yet Travis still runs on Ubuntu 14.04 which has version 9.3. Pointers on how to fix this are welcome.

**Comment by lerks** [comment]
I fixed the issue with PostgreSQL on Travis and the tests are passing now. Waiting for the review.

**Comment by stefano-maggiolo** [comment]
Remind me, why task and score type parameters are arrays of jsonb instead of jsonb?

---

Reviewed 8 of 17 files at r1, 7 of 9 files at r9, 6 of 8 files at r10, 6 of 6 files at r11, 1 of 1 files at r12, 2 of 4 files at r13, 2 of 2 files at r14, 2 of 2 files at r16, 1 of 1 files at r17, 1 of 1 files at r18, 6 of 6 files at r19.
Review status: all files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cms/db/base.py, line 109 at r15](https://reviewable.io:443/reviews/cms-dev/cms/741#-Krbvbk4hxhKcFglXcLM:-Krbvbk4hxhKcFglXcLN:b-3tmtkw) ([raw file](https://github.com/cms-dev/cms/blob/b505bae67da91e7f1c45481d3cc00d32e313dee3/cms/db/base.py#L109)):*
> ```Python
> 
>                 # Check that we understand the type
>                 if not isinstance(col.type, tuple(_TYPE_MAP.iterkeys())):
> ```

Given we do this quite often, shall we declare the tuple to a constant just after declaring _TYPE_MAP?

---

*[cms/db/base.py, line 220 at r15](https://reviewable.io:443/reviews/cms-dev/cms/741#-Krbw4eDOilRik1Q9bYF:-Krbw4eDOilRik1Q9bYG:b-w21om8) ([raw file](https://github.com/cms-dev/cms/blob/b505bae67da91e7f1c45481d3cc00d32e313dee3/cms/db/base.py#L220)):*
> ```Python
>                         py_item_type = tuple(
>                             v for k, v in _TYPE_MAP.iteritems()
>                             if isinstance(col.type.item_type, k))[0]
> ```

Extract to something like get_candidate_types(col)?

---

*[cms/db/base.py, line 209 at r19](https://reviewable.io:443/reviews/cms-dev/cms/741#-KmPvJSd4E2Js6F5kX8N:-KmPvJSd4E2Js6F5kX8O:b-u8gn4l) ([raw file](https://github.com/cms-dev/cms/blob/65ba88a534c76d4efb0131ee3831011863504e02/cms/db/base.py#L209)):*
> ```Python
>                     # TODO col.type.python_type contains the type that
>                     # SQLAlchemy thinks is more appropriate. We could
>                     # use that and drop _TYPE_MAP...
> ```

Is this comment (and the same one later) still appropriate?

---

*[cms/db/base.py, line 348 at r19](https://reviewable.io:443/reviews/cms-dev/cms/741#-KmPvTtklRVxqCWlkjCl:-KmPvTtklRVxqCWlkjCm:b-upwlz7) ([raw file](https://github.com/cms-dev/cms/blob/65ba88a534c76d4efb0131ee3831011863504e02/cms/db/base.py#L348)):*
> ```Python
>                                     "for keyword argument '%s', which requires "
>                                     "a list of '%s'"
>                                     % (type(item), prp.key, py_item_type))
> ```

Maybe also extract this to a common method now that the logic is more complex?

---

*[cmscontrib/DumpExporter.py, line 124 at r19](https://reviewable.io:443/reviews/cms-dev/cms/741#-KmPzAkmxiBQ-4nyjk-p:-KmPzAkmxiBQ-4nyjk-q:bhcvs1t) ([raw file](https://github.com/cms-dev/cms/blob/65ba88a534c76d4efb0131ee3831011863504e02/cmscontrib/DumpExporter.py#L124)):*
> ```Python
>         return value
>     elif isinstance(type_, String):
>         return value.decode('latin1')
> ```

I know this wasn't your code, but why not utf8?

---

*[cmscontrib/DumpExporter.py, line 132 at r19](https://reviewable.io:443/reviews/cms-dev/cms/741#-KmPzIjJoWKzFwigTgax:-KmPzIjJoWKzFwigTgay:b-5h1570) ([raw file](https://github.com/cms-dev/cms/blob/65ba88a534c76d4efb0131ee3831011863504e02/cmscontrib/DumpExporter.py#L132)):*
> ```Python
>         return list(encode_value(type_.item_type, item) for item in value)
>     elif isinstance(type_, CIDR):
>         return "%s" % value
> ```

str(value)

---

*[cmscontrib/DumpImporter.py, line 111 at r19](https://reviewable.io:443/reviews/cms-dev/cms/741#-KmPzQEZTaBjrsuxM4hK:-KmPzQEZTaBjrsuxM4hL:b-gsac8b) ([raw file](https://github.com/cms-dev/cms/blob/65ba88a534c76d4efb0131ee3831011863504e02/cmscontrib/DumpImporter.py#L111)):*
> ```Python
>         return value
>     elif isinstance(type_, String):
>         return value.encode('latin1')
> ```

Same question here

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/741#-:-Krbxag7Ggoor81D_gt2:b2gieoj)*

**Comment by stefano-maggiolo** [comment]
In general I don't like too much enforcing that they are arrays - it should be up to the score type to decide the format - also there is plenty of space to screw up the structure inside the array, so I don't see the point in enforcing the top-level.

One reason to let you do this would be the transition to "all score types are group score types", in that case the top-level would always be an array because that's enforced by the score type... Discuss.

---

Review status: 0 of 69 files reviewed at latest revision, 6 unresolved discussions, some commit checks failed.

---

*[cms/db/base.py, line 109 at r15](https://reviewable.io:443/reviews/cms-dev/cms/741#-Krbvbk4hxhKcFglXcLM:-L3sXg0aB8Y403eAU0jP:bg4mn6q) ([raw file](https://github.com/cms-dev/cms/blob/b505bae67da91e7f1c45481d3cc00d32e313dee3/cms/db/base.py#L109)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

We don't do it quite often, only at startup when the classes are declared. I think the performance gain is minimal and keeping it this way helps readability.
</blockquote></details>

Ok, I thought otherwise - I see you changed to a map lookup anyway, so winwin.

---

*[cms/grading/ScoreType.py, line 145 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sY_BjTtf81J88S_EY:-L3sY_BjTtf81J88S_EZ:b-56lqft) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cms/grading/ScoreType.py#L145)):*
> ```Python
> 
>         returns (float, str, float, str, [str]): respectively: the
>             score, the opaque data with additional information (e.g.
> ```

State the opaque data is in JSON

---

*[cms/server/admin/handlers/user.py, line 76 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sYxDncKGs2zi3pclF:-L3sYxDncKGs2zi3pclG:b-4hm0y8) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cms/server/admin/handlers/user.py#L76)):*
> ```Python
>             self.get_password(attrs, user.password, False)
> 
>             self.get_string(attrs, "email", empty=None)
> ```

Correct to have removed preferred_languages?

---

*[cmscontrib/AddUser.py, line 66 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sZO6MyjZW8OprS6JD:-L3sZO6MyjZW8OprS6JE:b-5nnhz5) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cmscontrib/AddUser.py#L66)):*
> ```Python
>         preferred_languages = \
>             "[" + ",".join("\"" + lang + "\""
>                            for lang in preferred_languages.split(",")) + "]"
> ```

Why not `json.dumps(preferred_languages.split(","))`?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/741#-:-L3sWrbHQRitGjKIAEf9:bw3ffas)*

File: cms/db/base.py [hunk 1]
```diff
@@ -24,6 +24,7 @@
 from __future__ import print_function
 from __future__ import unicode_literals
 
+import ipaddress
```
Comment (stefano-maggiolo) [inline cms/db/base.py]: Better commit message please (Improve how?)

File: cms/db/submission.py [hunk 2]
```diff
@@ -711,11 +714,14 @@ class Evaluation(Base):
...
-    # String containing output from the grader (usually "Correct",
-    # "Time limit", ...).
-    text = Column(
-        String,
-        nullable=True)
+    # The output from the grader, usually "Correct", "Time limit", ...
+    # (to allow localization the first item of the list is a format
+    # string, possibly containing some "%s", that will be filled in
+    # using the remaining items of the list).
+    execution_text = Column(
```
Comment (stefano-maggiolo) [inline cms/db/submission.py]: I wouldn't change the name without a good reason

File: cmscontrib/updaters/update_29.py [hunk 3]
```diff
@@ -35,6 +35,14 @@
 logger = logging.getLogger(__name__)
 
 
+def fix_text(t):
+    if t is None:
+        return []
+    t = json.loads(t)
```
Comment (stefano-maggiolo) [inline cmscontrib/updaters/update_29.py]: It is possible t was a single string and not a JSON-encoded array - so if loads fails you need to return [t].

**Comment by lerks** [comment]
Review status: 0 of 69 files reviewed at latest revision, 10 unresolved discussions, some commit checks failed.

---

*[cms/server/admin/handlers/user.py, line 76 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sYxDncKGs2zi3pclF:-L3tuIYnsjtM-cW5ST3A:b-spimwl) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cms/server/admin/handlers/user.py#L76)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Correct to have removed preferred_languages?
</blockquote></details>

What do you mean? The call for `preferred_languages` is still there, a couple of lines down, and it has become a call to `get_string_list` (rather than `get_string`).

---

*[cmscontrib/AddUser.py, line 66 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sZO6MyjZW8OprS6JD:-L3tuqgOWGUd15XnUZ2t:biukxot) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cmscontrib/AddUser.py#L66)):*
<details><summary><i>Previously, stefano-maggiolo (Stefano Maggiolo) wrote…</i></summary><blockquote>

Why not `json.dumps(preferred_languages.split(","))`?
</blockquote></details>

Again, not sure what you mean. We need to parse a string containing a comma-separated list into a Python list. How would dumping it to JSON help?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/741)*

File: cmscontrib/updaters/update_29.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python2
```
[the remark this answers is not in the document]
Comment (lerks) [reply cmscontrib/updaters/update_29.py]: A RepeatedUnicode was already dumped as a list of strings, so there is no need to adapt it.

File: cmscontrib/updaters/update_29.py [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+"""
+
+from __future__ import absolute_import
+from __future__ import unicode_literals
+from __future__ import print_function
+
+import json
+import logging
+
```
[the remark this answers is not in the document]
Comment (lerks) [reply cmscontrib/updaters/update_29.py]: Done.

File: cmscontrib/updaters/update_29.py [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+            if k.startswith("_"):
+                continue
+
+            if v["_class"] == "Dataset":
+                v["task_type_parameters"] = \
+                    json.loads(v["task_type_parameters"])
+                v["score_type_parameters"] = \
+                    json.loads(v["score_type_parameters"])
+                if v["score_type"] == "Sum":
+                    v["score_type_parameters"] = [v["score_type_parameters"]]
```
[the remark this answers is not in the document]
Comment (lerks) [reply cmscontrib/updaters/update_29.py]: Done.

**Comment by stefano-maggiolo** [comment]
Review status: 0 of 63 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[cms/server/admin/handlers/user.py, line 76 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sYxDncKGs2zi3pclF:-L3vuippl2lN-T8zvyJe:biwhrzz) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cms/server/admin/handlers/user.py#L76)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

What do you mean? The call for `preferred_languages` is still there, a couple of lines down, and it has become a call to `get_string_list` (rather than `get_string`).
</blockquote></details>

Reviewable was getting confused with all the rebases, sorry

---

*[cmscontrib/AddUser.py, line 66 at r20](https://reviewable.io:443/reviews/cms-dev/cms/741#-L3sZO6MyjZW8OprS6JD:-L3vv1ySt_-ulrERvdrn:brs86ps) ([raw file](https://github.com/cms-dev/cms/blob/858d6cc041595b8f113948314ead287ad7244bed/cmscontrib/AddUser.py#L66)):*
<details><summary><i>Previously, lerks (Luca Wehrstedt) wrote…</i></summary><blockquote>

Again, not sure what you mean. We need to parse a string containing a comma-separated list into a Python list. How would dumping it to JSON help?
</blockquote></details>

Again, for some reason reviewable is showing me the diff in the other direction -.-
Ignore.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cms-dev/cms/741)*

**Comment by stefano-maggiolo** [comment]
Ok for everything but that comment on fix_text: I think the text could be either a single string or a JSON array of strings, so you may want to revert to return [t] if json.loads fails

**Pull request closed without merging**